pallets/pallet-migration/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2023 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use crate::default_weights::WeightInfo;
+pub use pallet::*;
+
+pub mod default_weights;
+
+#[cfg(feature = "runtime-benchmarks")]
+mod benchmarking;
+#[cfg(any(test, feature = "runtime-benchmarks"))]
+mod mock;
+#[cfg(test)]
+mod test;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use attestation::{Attestations, ClaimHashOf};
+	use core::fmt::Debug;
+	use delegation::{DelegationNodeIdOf, DelegationNodes};
+	use did::{Did, DidIdentifierOf};
+	use frame_support::{
+		pallet_prelude::*,
+		traits::{fungible::Inspect, Currency, ReservableCurrency},
+	};
+	use frame_system::pallet_prelude::*;
+	use kilt_support::traits::MigrationManager;
+	use pallet_did_lookup::{linkable_account::LinkableAccountId, ConnectedDids};
+	use pallet_web3_names::{Owner, Web3NameOf};
+	use public_credentials::{CredentialIdOf, Credentials, SubjectIdOf};
+	use sp_runtime::traits::Hash;
+	use sp_runtime::SaturatedConversion;
+
+	use crate::default_weights::WeightInfo;
+
+	pub(crate) type AccountIdOf<T> = <T as frame_system::Config>::AccountId;
+
+	pub(crate) type BalanceOf<T> = <<T as Config>::Currency as Currency<AccountIdOf<T>>>::Balance;
+
+	pub type HashOf<T> = <T as frame_system::Config>::Hash;
+
+	#[derive(Encode, Decode, TypeInfo, Debug, Clone, PartialEq)]
+	pub struct EntriesToMigrate<T>
```

Comment:
If we lazied migrate the staking pallet, there would be a period of time when we had no collators. So no block authoring until some have migrated their locks into freezes. Currently, there are less than 30 collators and not many delegators. The direct migration is quite cheap.

Reply:
I see that the migration is quite cheap, so that is fine. But wouldn't we simply have to adapt the logic for choosing collators so that it considers both holds and reserves? Why do you say we would have no collators?

Reply:
Yes, sure. I just wanted to understand it a bit better 😊